picus/algorithms/dpvl.rkt

```diff
@@ -260,66 +260,75 @@
+      (cons (format "~a" (hash-ref r2c-map (number->string k))) val)))
   (map process-subinfo info))
 
-;; partition-vars :: (or/c '() hash?), set?, set? -> (list/c hash? hash? hash? hash?)
+;; partition-vars :: (or/c '() hash?), set?, set? -> (list/c hash? hash? hash? hash? hash?)
 (define (partition-vars info input-set output-set)
   (define pinfo (if (list? info) (make-hash) info)) ; patch for info type, fix later
-  (for/fold ([input-info (hash)]
-             [output1-info (hash)]
-             [output2-info (hash)]
-             [other-info (hash)]
-             #:result (list input-info output1-info output2-info other-info))
+  (for/fold ([input-info '()]
+             [output1-info '()]
+             [output2-info '()]
+             [other1-info '()]
+             [other2-info '()]
+             #:result (list input-info output1-info output2-info other1-info other2-info))
             ([(k v) (in-hash pinfo)])
     (match k
       ;; matching for a variable in the input set. By construction,
       ;; this variable is in the form x<number-id>.
       [(pregexp #px"^x(\\d+)$" (list _ (app string->number n)))
        #:when (set-member? input-set n)
-       (values (hash-set input-info k v)
+       (values (cons (cons n v) input-info)
                output1-info
                output2-info
-               other-info)]
+               other1-info
+               other2-info)]
       ;; matching for a variable in the output set. By construction,
       ;; this variable is in the form x<number-id> or y<number-id>,
       ;; where x indicates that it is in the first set
       ;; and y indicates that it is in the second set.
       [(pregexp #px"^(?:x|y)(\\d+)$" (list _ (app string->number n)))
        #:when (set-member? output-set n)
        (values input-info
-               (if (string-prefix? k "x") (hash-set output1-info k v) output1-info)
-               (if (string-prefix? k "y") (hash-set output2-info k v) output2-info)
-               other-info)]
-      [_ (values input-info
-                 output1-info
-                 output2-info
-                 (hash-set other-info k v))])))
+               (if (string-prefix? k "x") (cons (cons n v) output1-info) output1-info)
+               (if (string-prefix? k "y") (cons (cons n v) output2-info) output2-info)
+               other1-info
+               other2-info)]
+      ;; matching for internal signals. By construction,
+      ;; this variable is in the form x<number-id> or y<number-id>,
+      ;; where x indicates that it is in the first set
+      ;; and y indicates that it is in the second set.
+      [(pregexp #px"^(?:x|y)(\\d+)$" (list _ (app string->number n)))
+       (values input-info
+               output1-info
+               output2-info
+               (if (string-prefix? k "x") (cons (cons n v) other1-info) other1-info)
+               (if (string-prefix? k "y") (cons (cons n v) other2-info) other2-info))]
+
+      [(pregexp #px"^(?:ps\\d|zero|one|p)$")
```

Comment:
What is this regex looking for?

Reply:
It looks for symbolic variables that are inserted by Picus, but not defined by the circuit itself. I can add a comment explaining that.

Reply:
That would be great thanks!